controllers/topology/internal/extensions/patches/engine.go

```diff
@@ -0,0 +1,495 @@
+	// Return a JSONPatchGenerator if there are PatchDefinitions in the patch.
+	if len(patch.Definitions) > 0 {
+		// TODO(sbueringer): will be implemented in a follow-up PR
+		// return inline.NewJSONPatchGenerator(patch), nil
+		return nil, nil
+	}
+
+	return nil, errors.Errorf("failed to create patch generator for patch %q", patch.Name)
+}
+
+// applyPatchesToRequest updates the templates of a GenerateRequest by applying the patches
+// of a GenerateResponse.
+func applyPatchesToRequest(ctx context.Context, req *api.GenerateRequest, resp *api.GenerateResponse) error {
+	log := tlog.LoggerFrom(ctx)
+
+	for _, patch := range resp.Items {
+		log = log.WithValues("templateID", templateIDRef(patch.TemplateID))
+
+		// Get the template the patch should be applied to.
+		template := getTemplate(req, patch.TemplateID)
+
+		// If a patch doesn't apply to any template, this is a misconfiguration.
+		if template == nil {
+			return errors.Errorf("could not find template with id %q in GenerateRequest", templateIDRef(patch.TemplateID))
+		}
+
+		// Use the patch to create a patched copy of the template.
+		var patchedTemplate []byte
+		var err error
+
+		switch patch.PatchType {
+		case api.JSONPatchType:
+			log.V(5).Infof("Accumulating JSON patch", "patch", string(patch.Patch.Raw))
+			jsonPatch, err := jsonpatch.DecodePatch(patch.Patch.Raw)
+			if err != nil {
+				return errors.Wrapf(err, "failed to apply patch to template %s: error decoding json patch (RFC6902): %s",
+					templateIDRef(template.TemplateID), string(patch.Patch.Raw))
+			}
+
+			patchedTemplate, err = jsonPatch.Apply(template.Template.Raw)
+			if err != nil {
+				return errors.Wrapf(err, "failed to apply patch to template %s: error applying json patch (RFC6902): %s",
+					templateIDRef(template.TemplateID), string(patch.Patch.Raw))
+			}
+		case api.JSONMergePatchType:
+			log.V(5).Infof("Accumulating JSON merge patch", "patch", string(patch.Patch.Raw))
+			patchedTemplate, err = jsonpatch.MergePatch(template.Template.Raw, patch.Patch.Raw)
+			if err != nil {
+				return errors.Wrapf(err, "failed to apply patch to template %s: error applying json merge patch (RFC7386): %s",
+					templateIDRef(template.TemplateID), string(patch.Patch.Raw))
+			}
+		}
+
+		// Overwrite the spec of template.Template with the spec of the patchedTemplate,
+		// to ensure that we only pick up changes to the spec.
+		if err := patchTemplateSpec(&template.Template, patchedTemplate); err != nil {
+			return errors.Wrapf(err, "failed to apply patch to template %s",
+				templateIDRef(template.TemplateID))
+		}
+	}
+	return nil
+}
+
+// updateDesiredState uses the patched templates of a GenerateRequest to update the desired state.
+// NOTE: This func should be called after all the patches have been applied to the GenerateRequest.
+// TODO(sbueringer): Let's only overwrite desired state objects if the corresponding template actually have been patched.
```

Comment:
follow up PR?

Reply:
Given that it's an optimization and I'm not sure if the current PR is already to big, I would say yes.

Reply:
Looks like I can fix it when implementing the correct diff.